### ISSUE TITLE
Fix regression: double encoding of option values

### ISSUE THREAD

**Issue by laryn** [backdrop/backdrop-issues#6126]
## Description of the bug

I noticed this on a views filter for a date field, which has a `>=` in the option value. Improvements that came in with [this commit](https://github.com/backdrop/backdrop/commit/5d202e38f2eca1714055505a4675a61f726eb4c6) accidentally introduced a double-encoding of the option values. In this case, that means `>=` becomes `&gt;=` and no longer validates.

Within `form_select_options` the key value is run through `check_plain` and then used in `backdrop_attributes`, which runs the value through `check_plain` again. 

PR coming.

**Comment by kswan** [comment]
This looks similar to a problem @cellear mentioned in Zulip. https://backdrop.zulipchat.com/#narrow/stream/218635-Backdrop/topic/1.2E25.2E0.20release/near/358362045

I tested @cellear's case in the PR sandbox and it isn't fixed so there may be separate issues.

**Comment by argiepiano** [comment]
Tested and reviewed. RTBC!

**Comment by klonos** [comment]
I have also tested, to confirm that the removal of `check_plain()` doesn't allow things like HTML tags to "bleed" into the options:

![image](https://github.com/backdrop/backdrop-issues/assets/2423362/4d4c2a2a-75dd-4fa7-80eb-62dea74ff7c2)

![image](https://github.com/backdrop/backdrop-issues/assets/2423362/c50a5b5d-8086-4017-b612-8a3af09946c6)

**Comment by kiamlaluno** [comment]
Should not `nasty option 2 with <script>` be rendered as *nasty option 2 with &lt;script&gt;*, where `<` and `>` are escaped?

**Comment by klonos** [comment]
@kiamlaluno my understanding was that the entire HTML tag would be stripped. However, I have tested the same on a demo Sandbox with current v1.25.0 of core, and it behaves the same/similarly:
- in 1.25.0, the option is being rendered like so:
  ```html
  <option value="nasty option 2 with &amp;lt;script&amp;gt;">nasty option 2 with </option>
  ```
- in the PR sandbox, the option is rendered like so:
  ```html
  <option value="nasty option 2 with <script>">nasty option 2 with </option>
  ```

So in both cases it seems that the browser (tested both only in Chrome) is simply not rendering the `&amp;lt;script&amp;gt;` or `<script>`.

@laryn I'm thinking that perhaps we should be running it through `filter_xss()` instead 🤔 ...I'll test it quickly on my local.

**Comment by klonos** [comment]
...I should point out that within the same `else` that was changed, I should have looked at how the final output is being constructed, which is this:

```php
$options .= '<option' . backdrop_attributes($option_attributes) . '>' . check_plain($choice) . '</option>';
```

So the visible part that I have checked (which is the `$choice` var) is still being run through `check_plain()` anyway.

Anyway, I have checked on my local with `nasty option with <script> and something <strong>strong</strong>` as the option (which sets the same for both the key and the value, so it should end up in both the option `value` attribute, as well as its label) and with `filter_xss()` instead, and this is what happens:
- without any check (as in the PR):
  ```html
    <option value="nasty option with <script> and something <strong>strong</strong>">nasty option with  and something strong</option>
  ````
- with `check_plain()`, as currently in 1.x:
  ```html
    <option value="nasty option with &amp;lt;script&amp;gt; and something &amp;lt;strong&amp;gt;strong&amp;lt;/strong&amp;gt;">nasty option with  and something strong</option>
  ````
- with `filter_xss()` or `filter_xss_admin()`:
  ```html
    <option value="nasty option with  and something <strong>strong</strong>">nasty option with  and something strong</option>
  ````
- with good-ol' `strip_tags()`:
  ```html
    <option value="nasty option with  and something strong">nasty option with  and something strong</option>
  ````

So now I'm wondering whether we should be simply running it via `strip_tags()` (because HTML tags don't belong in `value` attributes) 🤔

**Comment by kiamlaluno** [comment]
The `&amp;lt;` is `<` escaped twice; that is the issue.

HTML tags are not removed from plain text added to HTML markup (the page content); they are escaped (once). That is what Backdrop does for the page title, for example.

```php
function backdrop_get_title() {
  $title = backdrop_set_title();

  // During a bootstrap, menu.inc is not included and thus we cannot provide a title.
  if (!isset($title) && function_exists('menu_get_active_title')) {
    $title = check_plain(menu_get_active_title());
  }

  return $title;
}
```

[`backdrop_get_title()`](https://docs.backdropcms.org/api/backdrop/core%21includes%21bootstrap.inc/function/backdrop_get_title/1) does not use `strip_tags()`. It just needs to avoid that rendering a title like `My name is <script type = "text/javascript">alert("Darth Vader")</script>` executes the JavaScript code; for that is sufficient to change the text to `&lt;script type = &quot;text/javascript&quot;&gt;alert(&quot;Darth Vader&quot;)&lt;/script&gt;` which is what `check_plain()` does. (Technically, to avoid that the JavaScript code is run is sufficient to escape `<` and `>`; escaping the quote characters avoid issue with `<span title="<script type = "text/javascript">alert("Darth Vader")</script>" >example</span>`, for example.)

**Comment by kiamlaluno** [comment]
See also what [`backdrop_attributes()`](https://docs.backdropcms.org/api/backdrop/core%21includes%21common.inc/function/backdrop_attributes/1) does to avoid that, for example, a *title* attribute set to `<script>steal_cookie();</script>` causes the JavaScript code to be executed.

```php
function backdrop_attributes(array $attributes = array()) {
  foreach ($attributes as $attribute => &$data) {
    $data = implode(' ', (array) $data);
    $data = $attribute . '="' . check_plain($data) . '"';
  }
  return $attributes ? ' ' . implode(' ', $attributes) : '';
}
```

It does not uses `strip_tags()`, but `check_plain()`.

**Comment by kiamlaluno** [comment]
Anyway, the key/label values for *Allowed values list* are not influenced by removing the `check_plain()` call because, for those values, the HTML tags that are not allowed are stripped out, as the description in the page says.

> Allowed HTML tags in labels: <a> <b> <big> <code> <del> <em> <i> <ins> <pre> <q> <small> <span> <strong> <sub> <sup> <tt> <ol> <ul> <li> <p> <br> <img>

**Comment by kiamlaluno** [comment]
On the preview site for the merge request, I created a *Status* for the *Page* content type, using the following key/label values.

- `Small`: `<small>Small text</small>`
- `Code`: `<code>$i = 10;</code>`
- `Bold`: `Be <strong>bold</strong>`

I chose *Checkboxes/radio buttons* as widget type.

When I create a *Page* node, the *Status* field is rendered as in this screenshot.

![screenshot](https://github.com/backdrop/backdrop-issues/assets/29851/66a642fe-bdd9-49eb-b7b6-26f1f98bd8e8)

Labels are correctly shown; no HTML tag is double-encoded.

**Comment by klonos** [comment]
@kiamlaluno yes. The point I was trying to make though was that HTML tags do not belong in the value the `value` attribute. So we should be stripping them.

https://www.w3.org/TR/2011/WD-html5-20110525/elements.html says the following:
> ## 3.2.4.1 Attributes
> Except where otherwise specified, attributes on HTML elements may have any string value, including the empty string. Except where explicitly stated, there is no restriction on what text can be specified in such attributes.

However it does also state that:
>Authors must not use elements, attributes, or attribute values for purposes other than their appropriate intended semantic purpose. ...

Also all elements listed in https://www.w3.org/TR/html4/index/attributes.html that allow for a `value` attribute are to have `CDATA` for its value (with the exception of `li` elements, that are to have `NUMBER`). And the definitions that I have found for `CDATA` are:
- https://www.w3schools.com/xml/dom_cdatasection.asp
  > A CDATA section contains text that will NOT be parsed by a parser. Tags inside a CDATA section will NOT be treated as markup and entities will not be expanded. The primary purpose is for including material such as XML fragments, without needing to escape all the delimiters.
- https://en.wikipedia.org/wiki/CDATA
  > ... The term indicates that a certain portion of the document is general character data, rather than non-character data or character data with a more specific, limited structure.
  > ...
  > In an XML document or external entity, a CDATA section is a piece of element content that is marked up to be interpreted literally, as textual data, not as marked up content. A CDATA section is merely an alternative syntax for expressing character data; there is no semantic difference between character data in a CDATA section and character data in standard syntax where, for example, `<` and `&` are represented by `&lt;` and `&amp;`, respectively.
- https://www.w3.org/TR/REC-xml/#sec-cdata-sect:
  > Definition: **CDATA sections** may occur anywhere character data may occur; they are used to escape blocks of text containing characters which would otherwise be recognized as markup. ...

Having said all that, we are using `backdrop_attributes()` for attributes other than `value`, which means that if we are to do anything with the string that is to be passed to the `value` attribute as its value, then we need to do it before we pass it to `backdrop_attributes()`.

**Comment by klonos** [comment]
@laryn based on my research and findings above, I'm setting this to NW, since passing the value as is is obviously wrong. We have the following options, which as I said in my previous comment need to happen before we pass the string to `backdrop_attributes()`:
- `check_plain()` which "decodes" HTML starting characters (`<`, `>` etc.) to `&amp;lt;` and `&amp;gt;` (but that one has the problem we are facing here, since `backdrop_attributes()` then also `check_plain()`'s it)
- `filter_xss()` or `filter_xss_admin()` cannot be used IMO, since they allow HTML.
- `strip_tags()` which basically removes "the nasties", and has the value be in CDATA format.

**Comment by klonos** [comment]
...I've tested on my local, and when using `strip_tags()`, the result is this:
```html
<option value="views filter with >=">views filter with &gt;=</option>
```

**Comment by laryn** [comment]
@klonos Here's a config archive that you should be able to import via the UI on a fresh sandbox (it's from the current sandbox for this issue). It sets up a date field on Posts and a simple View to test the date filter (>=).

[issue-6126-config.tar.gz](https://github.com/backdrop/backdrop-issues/files/11605390/issue-6126-config.tar.gz)

**Comment by klonos** [comment]
@laryn unfortunately I couldn't reproduce the problem. Here's what I did:
- span up a demo site via https://backdropcms.org/demo (that's core v1.25.0, on PHP v8.1.19) - link to the demo site + credentials [here](https://backdropcms.org/demo/64779feeeebad9c03d0dcedc)
- navigated to `admin/config/development/logging` and enabled showing all errors
- navigated to `admin/config/development/configuration/full/import` and imported the config export you have provided
- navigated to `admin/structure/views/view/date_test` and clicked the "Content: Date (now)" link in the filters section to edit the date field filter -> no errors 🤷🏼 

Am I missing any step?

**Comment by laryn** [comment]
@klonos Try to apply the filter after your last step (and notice that there are no configuration fields available for the "greater than or equal to" setting, which should be off to the right):

<img width="1424" alt="image" src="https://github.com/backdrop/backdrop-issues/assets/871421/c14b8e19-19c3-451b-9f51-8ccb24555ba5">

**Comment by klonos** [comment]
OK, got it now 👍🏼 ...thanks @laryn 🙏🏼 

Doing some more testing on my local as well.

**Comment by klonos** [comment]
OK, here's the output of that select element on vanilla 1.x (using `check_plain()`):

```html
<select id="edit-options-operator" name="options[operator]" class="form-select">
  <option value="&amp;lt;">Is less than</option>
  <option value="&amp;lt;=">Is less than or equal to</option>
  <option value="=" selected="selected">Is equal to</option>
  <option value="!=">Is not equal to</option>
  <option value="&amp;gt;=">Is greater than or equal to</option>
  <option value="&amp;gt;">Is greater than</option>
  <option value="between">Is between</option>
  <option value="not between">Is not between</option>
  <option value="empty">Is empty (NULL)</option>
  <option value="not empty">Is not empty (NOT NULL)</option>
  <option value="regular_expression">Regular expression</option>
  <option value="contains">Contains</option>
</select>
```

With the approach in the PR, things are fixed:

```html
<select id="edit-options-operator" name="options[operator]" class="form-select">
  <option value="<">Is less than</option>
  <option value="<=">Is less than or equal to</option>
  <option value="=">Is equal to</option>
  <option value="!=">Is not equal to</option>
  <option value=">=" selected="selected">Is greater than or equal to</option>
  <option value=">">Is greater than</option>
  <option value="between">Is between</option>
  <option value="not between">Is not between</option>
  <option value="empty">Is empty (NULL)</option>
  <option value="not empty">Is not empty (NOT NULL)</option>
  <option value="regular_expression">Regular expression</option>
  <option value="contains">Contains</option>
</select>
```

...but as I explained in my previous comments here, we shouldn't be doing that, as it allows "nasties" and HTML tags to end up in the value of `value=""`.

Using `strip_tags()` on the other hand "fixes" the issue, and also avoids HTML tags ending up in the value of `value=""` (which according to the "semantic usage" of that attribute and my understanding of the HTML specs by W3C that I went through, is invalid/inappropriate). However, notice the HTML output (how the values for the "Is less than" and "Is less than or equal to" options have been stripped completely - not ideal at all 😞 🤔 ):

```html
<select id="edit-options-operator" name="options[operator]" class="form-select">
  <option value="">Is less than</option>
  <option value="">Is less than or equal to</option>
  <option value="=">Is equal to</option>
  <option value="!=">Is not equal to</option>
  <option value=">=" selected="selected">Is greater than or equal to</option>
  <option value=">">Is greater than</option>
  <option value="between">Is between</option>
  <option value="not between">Is not between</option>
  <option value="empty">Is empty (NULL)</option>
  <option value="not empty">Is not empty (NOT NULL)</option>
  <option value="regular_expression">Regular expression</option>
  <option value="contains">Contains</option>
</select>
```

So here are my thoughts:
- we certainly need to be making sure that the `$key` is not output as is, but...
- `check_plain()` is problematic
- `strip_tags()` is also problematic apparently
- ideally, all the values of the options in that dropdown would be textual instead of having symbols, but I suspect that that would be a breaking change (need to think more about it and test) ...or perhaps we'd need to do some additional work during validation to make that happen, but that would be the "safest" way to go in my opinion at the moment.

**Comment by laryn** [comment]
I agree that text values would be better than symbols if we can swing it. However, the commit that caused this regression went into 1.25.0. I think we need to prioritize getting a fix into 1.25.1 to get functionality back to what it was, even if it means filing a follow-up after that for potentially reworking things. 

We are not outputting `$key` as is anywhere in this PR, we're reducing the number of times it goes through `check_plain` to one.

**Comment by klonos** [comment]
> We are not outputting `$key` as is anywhere in this PR ...

Yup, agreed ...well, sorta. If you see my earlier comment [here](https://github.com/backdrop/backdrop-issues/issues/6126#issuecomment-1565944902), and follow the steps blow, you will see why I am having concerns:
- in the PR sandbox (or on your local, with the changes from the PR applied), either create a new content type, or edit an existing one, and head to the "Manage fields" page
- Add a new field of type "List (text)" (you can have the widget be either checkboxes/radios or a select - doesn't really matter)
- In the allowed values list for the field, add some "nasty" things (if you click the "Manual entry" option, make sure that the **keys** have some HTML tags etc. in them)
- Save the field and the content type
- Start creating a new node of that type, and inspect the HTML code of the field you just added
- Notice how HTML tags have ended up in the value of the `value=""` attribute 👎🏼

**Comment by klonos** [comment]
...perhaps a possible solution would be to work on some validation/sanitization in https://github.com/backdrop/backdrop/tree/1.x/core/modules/field/modules/options 🤔

**Comment by argiepiano** [comment]
>  I think we need to prioritize getting a fix into 1.25.1 to get functionality back to what it was, even if it means filing a follow-up after that for potentially reworking things.

I have to agree with @laryn here. While there is a potential risk (though this is really unclear, since html tag attributes are not rendered by the browser), I think the first priority is to get this to the state it was in before this regression, so that people can continue using common things like Views date and numerical filters.

Eventually, in a separate issue, we should consider running the value attribute through `check_plain`. To avoid potential security issues, I'd suggest that we re-code all instances where keys for select inputs might use special symbols like `<` and `<=` etc. This shouldn't be hard to do in the case of the Views numeric filters. After all the `short` versions of the operations (where the `<=` etc exists) are only used for displaying the hidden entry fields in the form, and for the admin summary.

**Comment by klonos** [comment]
...moving this back to RTBC. I will bring this up during the dev meeting in a few hours from now, and see if we have consensus.

**Comment by argiepiano** [comment]
Hello @laryn and @klonos. Wondering if this nasty bug will get fixed today? Was this discussed yesterday?

One more important datapoint here. After @laryn's fix of the regression (which basically sets the clock back to the way this was done before the regression), please notice the theme output that Backdrop is generating:

```html
<div class="form-item form-type-select form-item-options-operator">
  <label for="edit-options-operator">Operator </label>
  <select id="edit-options-operator" name="options[operator]" class="form-select">
    <option value="&lt;">Is less than</option>
    <option value="&lt;=">Is less than or equal to</option>
    <option value="=" selected="selected">Is equal to</option>
    <option value="!=">Is not equal to</option>
    <option value="&gt;=">Is greater than or equal to</option>
    <option value="&gt;">Is greater than</option>
    <option value="between">Is between</option>
    <option value="not between">Is not between</option>
    <option value="empty">Is empty (NULL)</option>
    <option value="not empty">Is not empty (NOT NULL)</option>
    <option value="regular_expression">Regular expression</option></select>
</div>
```

You will notice above that the `<` etc are all **correctly encoded** to `&lt;` etc. **This is what Backdrop is sending to the browser**. The browsers (all of them AFAICT) are actually decoding those entities back into `<` etc.   In short, I believe we are being overcautious with all this discussion. If this was a security risk, it would have been caught a long time ago and fixed in Drupal's world.

**Comment by klonos** [comment]
> Was this discussed yesterday?

Yes, we did discuss it, however we did not have @quicksketch in the call. I think that we should aim to merge this (and #3492, which was also mentioned in the dev meeting), and get a 1.25.1 bug fix release out over the weekend if possible (we can skip the rest of the RTBC bug fixes that are not for 1.25.x specifically). 

Pinging all our @backdrop/core-committers re that ^^

> If this was a security risk, it would have been caught a long time ago and fixed in Drupal's world.

I don't want to have to rely on that. Besides, I'm not worried about the HTML output in the browser, rather than what will be sent during submit (which is most likely what will be saved unless there is custom validation). I still think that we need either one or several follow-up issues to do the following:
- convert all the values of `value=""` in this specific select (in the Views UI) to be explicitly textual, then convert them into symbols in a form validation handler (so that we retain backwards compatibility and save what was previously being saved in the various views configs)
- see if there is any other instance in core where we are are adding odd/invalid and potentially dangerous values in the `value` attribute of various input elements (not just selects)
- special-case the `value` attribute in `backdrop_attributes()`, so that it only allows things specified in the HTML specs.
- see if other attributes need special handling, depending on what the specs say for them

**Comment by quicksketch** [comment]
The pull request looks good and matches what Backdrop/Drupal have been doing for decades regarding escaping values in select lists.

> special-case the value attribute in `backdrop_attributes()`, so that it only allows things specified in the HTML specs.

By escaping the value, everything is within HTML specifications. I don't think `value` needs to be any different from any other element attribute. For example we need to do the same escaping on `alt` attributes, where any kind of string is going to show up. As far as XSS problems are concerned, all HTML attributes share the same potential issues, which is why using `backdrop_attributes()` is a good idea. We just accidentally over-escaped the attribute in this change. I don't think any wide-scale changes are necessary, and it would be an API break to strip characters like `<` and `&` from option values.

**Comment by quicksketch** [comment]
I merged https://github.com/backdrop/backdrop/pull/4442 into 1.x for 1.25.1, which we'll want to get out as soon as possible. I'll shoot for later today or tomorrow.
